share/man/man1/gathero.1
.\" Manpage for gathero
.\" Please open an issue on GitHub or fork and push changes to the
.\" repository to correct errors or typos.

.TH gathero 1 "12 April 2021" "1.2.1" "gathero man page"
.SH NAME
gathero \- a script to collect essential information about a user's job(s).

.SH SYNOPSIS
gathero [-h] [-n NAME] [-c {gzip,bz2,xz,tar,zip}] [-d DIRECTORY] [-V] [--license] [jobs [jobs ...]]

.SH DESCRIPTION
gathero is an advanced, high\-level tool based of an extension of the checkjob utility available on most HPC clusters. Where gathero differs from checkjob is that it pulls information on both the user and allocation and then runs a series of commands on the user and the allocation. Once gathero is done, it saves the output to a directory that can be analyzed by the user. Gathero also creates are shareable archive as well and saves it to the output directory. Also, gathero has the ability to query multiple job ids. 

.SH OPTIONS
.IP "\fB-c|--compression\fP [gzip|bz2|xz|tar|zip]"
Compression algorithm to use when creating gathero archive. The default is zip compression.

.IP "\fB-d|--directory\fP <output_directory>"
Directory to save gathero\'s output directory to. On Roar, the default is the $HOME/scratch directory.

.IP "\fB-n|--name\fP <output_name>"
Name of gathero output directory. The default is gathero_output.

.IP "\fB-V|--version\fP"
Print gathero version info.

.IP "\fB--license\fP"
Print gathero licensing info.

.IP "\fB-h|--help\fP"
Print help info.

.SH FILES
.TP
.I
/gpfs/group/dml129/default/sw7/iaskutils/bin/gathero

.TP
.I
/gpfs/group/dml129/default/sw7/iaskutils/gathero.py

.TP
.I
/gpfs/group/dml129/default/sw7/iaskutils/utils/compression.py

.TP
.I
/gpfs/group/dml129/default/sw7/iaskutils/utils/print_license.py

.TP
.I
/gpfs/group/dml129/default/sw7/iaskutils/utils/special_print.py

.TP
.I
/gpfs/group/dml129/default/sw7/iaskutils/utils/verifylocale.py

.SH EXAMPLES
.TP
.BI "gathero 123456789"
.TP
.PP
Run gathero on job id 123456789 and save the output to $HOME/scratch/gathero_output. Creates a zip archive named gathero_output.zip.

.TP
.BI "gathero 123456789 987654321"
.TP
.PP
Run gathero on job ids 123456789 and 987654321 and save the output to $HOME/scratch/gathero_output. Creates a zip archive named gathero_output.zip.

.TP
.BI "gathero 123456789 987654321 -c gzip"
.TP
.PP
Run gathero on job ids 123456789 and 987654321 and save the output to $HOME/scratch/gathero_output. Creates a gzip archive named gathero_output.tar.gz.

.TP
.BI "gathero 123456789 987654321 -c xz -n jcn23_output"
.TP
.PP
Run gathero on job ids 123456789 and 987654321 and save the output to $HOME/scratch/jcn23_output. Creates an xz archive named jcn23_output.tar.xz.

.TP
.BI "gathero 123456789 -c bz2 -n jcn23_output -d ~/work"
.TP
.PP
Run gathero on job id 123456789 and save the output to $HOME/work/jcn23_output. Creates a bz2 archive named jcn23_output.tar.bz2.

.SH BUGS

The following is a list of known bugs in gathero:

.PP
.TP
.BI "gathero fails to execute correctly if gcc devtoolset is loaded on Roar:"
.TP
.PP
If you have the gcc module loaded while trying to use gathero, you may encounter an issue where gathero constantly asks for your sudo credentials. In order to avoid this issue, simply unload the gcc module before calling gathero.

.TP
.BI "If you encounter any other bugs using gathero, then please open an issue at https://github.com/ICDS-Roar/iaskutils."

.SH AUTHOR
Jason C. Nucciarone (jcn23@psu.edu, jason.nucciarone@gmail.com)

.SH COPYRIGHT
Copyright 2021 The Pennsylvania State University Institute for Computational and Data Sciences. All rights reserved at https://mit-license.org/.


gathero.py
import os
import subprocess
import multiprocessing
import shutil
import glob
from xml.dom import minidom
import argparse
from progress.bar import Bar
from utils.special_print import print_good
from utils.special_print import print_bad
from utils.special_print import print_info
from utils.special_print import important_info
from utils.special_print import printrichtext
from utils.compression import Compression
from utils.print_license import licenseheader
from utils.print_license import licensebody
from utils.verifylocale import verifylocale


# Global variables
home_env_var = os.getenv('HOME')


def checkjob(checkjob_path, job_id, root_dir):
    """Execute checkjob on the command line and then write to output file."""
    print_info("Running checkjob on {}".format(job_id))
    checkjob_out = subprocess.run(["sudo", "{}".format(checkjob_path), "{}".format(job_id), "-v", "--timeout=300"],
                                  capture_output=True, text=True)

    # Change into output directory and create output file
    os.chdir(root_dir)
    fout = open("{}_checkjob-output.txt".format(job_id), "wt")
    fout.write(str(checkjob_out.stdout))
    fout.close()


def checkjob_xml(checkjob_path, job_id, root_dir):
    """Execute checkjob on the command line, retrieve xml output, and then write to xml outfile."""
    checkjob_out = subprocess.run(["sudo", "{}".format(checkjob_path), "{}".format(job_id), "-v", "--timeout=300", "--xml"],
                                  capture_output=True, text=True)
    os.chdir(root_dir)
    fout = open("{}.xml".format(job_id), "wt")
    fout.write(str(checkjob_out.stdout))
    fout.close()


def gathero(job, name, compression, directory, version, license):
    """gathero: A script to collect essential information about a user's job(s)."""
    if version:
        licenseheader("gathero v1.2.1") 

    elif license:
        licensebody("gathero: A script to collect essential information about a user's job(s).")

    else:
        # Change into the directory specified by the user to create $NAME
        try:
            os.chdir(directory)

        except OSError:
            print_bad("Something went wrong when trying to change into {} ".format(directory) +
                      "\nPlease contact i-ASK center at iask@ics.psu.edu.")
            return

        # Create $NAME directory and then move into it.
        # Also save path above output dir because we
        # will need it later.
        dir_above_output = os.getcwd()
        try:
            os.mkdir(name)
            os.chdir(name)

        except FileExistsError:
            shutil.rmtree(name)
            os.mkdir(name)
            os.chdir(name)

        # Get system path to checkjob and mam-* executables
        checkjob_path = subprocess.getoutput("command -v checkjob")
        mam_list_funds_path = subprocess.getoutput("command -v mam-list-funds")
        mam_list_accounts_path = subprocess.getoutput("command -v mam-list-accounts")

        # If the user specified no job ids on the command line
        if job == []:
            printrichtext("No job ID found. Please enter job ID(s). \n"
                          "Use [bold blue]gathero --help[/bold blue] or contact the i-ASK center \n"
                          "at iask@ics.psu.edu if you need help.")
            return

        print_good("Executing checkjob on given job ids.")
        process_list = list()
        for jobs in job:
            try:
                process = multiprocessing.Process(target=checkjob,
                                                  args=(checkjob_path, jobs, os.getcwd()))
                process_list.append(process)
                process.start()

            except multiprocessing.ProcessError:
                print_bad("Something went wrong running checkjob. \n"
                          "Please contact i-ASK center for help!")

        # Block until all checkjobs have been completed
        for process in process_list:
            process.join()

        # Once checkjob has finished running
        print_good("All checkjobs have been successful!")
        print_good("Collecting allocation and user info.")
        
        # Run checkjob again, but this time pull the xml
        os.mkdir("tmp")
        above_tmp_dir = os.getcwd()
        os.chdir("tmp")

        process_list_2 = list()
        for jobs in job:
            try:
                process = multiprocessing.Process(target=checkjob_xml,
                                                  args=(checkjob_path, jobs, os.getcwd()))
                process_list_2.append(process)
                process.start()

            except multiprocessing.ProcessError:
                print_bad("Something went wrong collecting allocation and user info. \n"
                          "Please contact i-ASK center for help!")

        # Block until process 2 jobs are done
        for process in process_list_2:
            process.join()

        # After XML files have been collected, pull allocation name and user info
        xml_files = glob.glob("*.xml")
        alloc_name = list()
        user_id = list()

        for xml_file in xml_files:
            doc = minidom.parse(xml_file)
            data = doc.getElementsByTagName("job")

            for datum in data:
                if datum.attributes['Account'].value not in alloc_name:
                    alloc_name.append(datum.attributes['Account'].value)
                
                if datum.attributes['User'].value not in user_id:
                    user_id.append(datum.attributes['User'].value)

        # Process allocation and user info
        os.chdir(above_tmp_dir)
        shutil.rmtree("tmp")

        if len(alloc_name) != 0:
            for alloc in alloc_name:
                fout = open("{}-info.txt".format(alloc), "wt")
                # Run showq
                showq = subprocess.run(["showq", "-w", "acct={}".format(alloc)], 
                                        capture_output=True, text=True)
                fout.write("showq -w acct={}\n".format(alloc))
                fout.write(str(showq.stdout) + "\n")
                fout.close()

        if len(user_id) != 0:
            for user in user_id:
                fout = open("{}-info.txt".format(user), "wt")
                # Run account_quota_check
                account_quota_check = subprocess.run(["account_quota_check", "{}".format(user)], 
                                                        capture_output=True, text=True)
                fout.write("account_quota_check {}\n".format(user))
                fout.write(str(account_quota_check.stdout) + "\n\n")

                # Run qstat
                qstat = subprocess.run(["qstat", "-u", "{}".format(user)], 
                                        capture_output=True, text=True)
                fout.write("qstat -u {}\n".format(user))
                fout.write(str(qstat.stdout) + "\n\n")

                # Run mam-list-accounts
                mam_list_accounts = subprocess.run(["sudo", "{}".format(mam_list_accounts_path), "-u", "{}".format(user)],
                                                    capture_output=True, text=True)
                fout.write("mam-list-accounts -u {}\n".format(user))
                fout.write(str(mam_list_accounts.stdout) + "\n\n")

                # Run mam-list-funds
                mam_list_funds = subprocess.run(["sudo", "{}".format(mam_list_funds_path), "-u", "{}".format(user)],
                                                    capture_output=True, text=True)
                fout.write("mam-list-funds -u {}\n".format(user))
                fout.write(str(mam_list_funds.stdout))

        # Compress the gathero_output directory
        os.chdir(dir_above_output)
        compress_dir = above_tmp_dir.split("/")
        archive = Compression(compress_dir[-1], name, above_tmp_dir)

        if compression == "gzip":
            archive.togzip()
            shutil.move("{}/{}.tar.gz".format(dir_above_output, name), "{}/{}.tar.gz".format(above_tmp_dir, name))

        elif compression == "bz2":
            archive.tobzip()
            shutil.move("{}/{}.tar.bz2".format(dir_above_output, name), "{}/{}.tar.bz2".format(above_tmp_dir, name))

        elif compression == "xz":
            archive.toxz()
            shutil.move("{}/{}.tar.xz".format(dir_above_output, name), "{}/{}.tar.xz".format(above_tmp_dir, name))

        elif compression == "tar":
            archive.totar()
            shutil.move("{}/{}.tar".format(dir_above_output, name), "{}/{}.tar".format(above_tmp_dir, name))

        elif compression == "zip":
            archive.tozip()
            shutil.move("{}/{}.zip".format(dir_above_output, name), "{}/{}.zip".format(above_tmp_dir, name))

        else:
            print_bad("There was an issue compressing your files! Please contact the i-ASK center at " +
                      "iask@ics.psu.edu")
            return

        print_good("All done!")


if __name__ == "__main__":
    out, err = verifylocale()
    if err != None:
        printrichtext("Uh oh. Looks like the UTF-8 locale is not supported on your system! " +
                      "Please try using [bold blue]locale-gen en_US.UTF-8[/bold blue] before continuing.")
    
    else:
        parser = argparse.ArgumentParser()
        parser.add_argument("jobs", nargs="*")
        parser.add_argument("-n", "--name", default="gathero-output", help="Name of the output directory and archive (default: gathero-output).")
        parser.add_argument("-c", "--compression", type=str, choices=["gzip", "bz2", "xz", "tar", "zip"],
                            default="zip", help="Compression algorithm to use (default: zip).")
        parser.add_argument("-d", "--directory", type=str, default="{}/scratch".format(home_env_var), 
                            help="Directory to save output to (default: ~/scratch).")
        parser.add_argument("-V", "--version", action="store_true", help="Print version info.")
        parser.add_argument("--license", action="store_true", help="Print licensing info.")
        args = parser.parse_args()
        gathero(args.jobs, args.name, args.compression, args.directory, args.version, args.license)
